ocaml/typing/typetexp.ml

```diff
@@ -365,36 +425,73 @@ let newvar ?name layout =
+        raise Already_bound;
+      name
+  in
+  let ty = if generic
+    then Btype.newgenvar ~name layout
+    else
+      let ty = new_global_var ~name layout in
+      Option.iter (fun name -> TyVarEnv.add name ty) name_opt;
+      ty
+  in
+  { ctyp_desc = tvar; ctyp_type = ty; ctyp_env = env;
+    ctyp_loc = loc; ctyp_attributes = attrs }
+
+let transl_type_param_jst ~generic env loc attrs path :
+  Jane_syntax.Core_type.t -> _ =
+  function
+  | Jtyp_layout (Ltyp_var { name; layout = annot }) ->
+     let layout =
+       Layout.of_annotation ~context:(Type_parameter (path, name)) annot
+     in
+     transl_type_param_var ~generic env loc attrs name layout (Some annot.txt),
+     layout
+  | Jtyp_layout (Ltyp_poly _ | Ltyp_alias _) -> assert false
+
+let transl_type_param ~generic env path styp : Typedtree.core_type * layout =
   let loc = styp.ptyp_loc in
-  match styp.ptyp_desc with
-    Ptyp_any ->
-      let ty = new_global_var ~name:"_" layout in
-        { ctyp_desc = Ttyp_any; ctyp_type = ty; ctyp_env = env;
-          ctyp_loc = loc; ctyp_attributes = styp.ptyp_attributes; }
-  | Ptyp_var name ->
-      let ty =
-          if not (valid_tyvar_name name) then
-            raise (Error (loc, Env.empty, Invalid_variable_name ("'" ^ name)));
-          if TyVarEnv.is_in_scope name then
-            raise Already_bound;
-          let v = new_global_var ~name layout in
-          TyVarEnv.add name v;
-          v
-      in
-        { ctyp_desc = Ttyp_var name; ctyp_type = ty; ctyp_env = env;
-          ctyp_loc = loc; ctyp_attributes = styp.ptyp_attributes; }
-  | _ -> assert false
+  match Jane_syntax.Core_type.of_ast styp with
+  | Some (etyp, attrs) -> transl_type_param_jst ~generic env loc attrs path etyp
+  | None ->
+  (* Our choice for now is that if you want a parameter of layout any, you have
+   to ask for it with an annotation.  Some restriction here seems necessary
+   for backwards compatibility (e.g., we wouldn't want [type 'a id = 'a] to
+   have layout any).  But it might be possible to infer any in some cases. *)
+  let layout = Layout.of_new_sort_var ~why:Unannotated_type_parameter in
+  let attrs = styp.ptyp_attributes in
+  let cty =
+    match styp.ptyp_desc with
+      Ptyp_any -> transl_type_param_var ~generic env loc attrs None layout None
+    | Ptyp_var name ->
+      transl_type_param_var ~generic env loc attrs (Some name) layout None
+    | _ -> assert false
+  in
+  cty, layout
 
-let transl_type_param env styp layout =
+let transl_type_param ~generic env path styp =
   (* Currently useless, since type parameters cannot hold attributes
      (but this could easily be lifted in the future). *)
   Builtin_attributes.warning_scope styp.ptyp_attributes
-    (fun () -> transl_type_param env styp layout)
+    (fun () -> transl_type_param ~generic env path styp)
+
+(* returns just the layout of the param *)
+let transl_type_param_layout env path styp =
+  snd (transl_type_param ~generic:false env path styp)
```

Comment:
This seems wasteful - why translate the whole parameter and update the `TyVarEnv` just to throw that work away?  Can't this just be something like:
```ocaml
  match Jane_syntax.Core_type.of_ast styp with
  | None -> Layout.of_new_sort_var ~why:Unannotated_type_parameter
  | Some (Jtyp_layout (Ltyp_var { layout = annot; name }), _) ->
    Layout.of_annotation ~context:(Type_parameter (path, name)) annot
  | Some _ -> assert false
```
(And then a name like `get_type_param_layout` is probably more appropriate)